### DISCUSSION TITLE
Proposed handling of common Prisma errors

### DISCUSSION THREAD

**Discussion by coby111** [CADELmx/GPD_BACKEND#23] [Ideas]
## Context

In our CRUD methods, we use Prisma as an ORM to interact with the database. Currently, errors occurring during these operations are handled generically, which can make it difficult to identify and resolve specific problems. To improve the robustness and clarity of error responses in our application, I propose centralizing Prisma's error handling in a separate file. This approach will allow us to handle errors uniformly and consistently during CRUD operations.

## Identified Problems

1. **Generic error handling**: Errors are handled generically, returning a standard error message without providing specific details.
2. **User Experience**: Lack of specific error messages can confuse users and make troubleshooting difficult.
3. **Problem Diagnosis**: Generic error messages make it difficult to identify the root cause of database problems.

## Proposal

Implement specific error handling using Prisma known error codes (`PrismaClientKnownRequestError`). This will allow clearer and more detailed error messages to be returned to both users and developers.

### Common Prisma Error Codes

1. **P2000**: The value provided for the column is too long for the column type.
2. **P2001**: The requested record was not found.
3. **P2002**: Unique constraint failed.
4. **P2003**: Foreign key constraint failed.
5. **P2004**: Constraint check failed.
6. **P2005**: Invalid value.
7. **P2006**: Stored value is not valid.
8. **P2007**: The database returned an unexpected error (data validation error).
9. **P2008**: The database query returned an error (query engine error).
10. **P2009**: Data validation error.

For more details, refer to the [Prisma error reference documentation](https://www.prisma.io/docs/orm/reference/error-reference)

### Implementation of the proposal
**Error Handling File:** A file called `prisma-error-handler.ts` can be created in a suitable directory (for example, `src/common/`).

```ts
import { PrismaClientKnownRequestError } from '@prisma/client/runtime';
import { NotFoundException, BadRequestException, ConflictException } from '@nestjs/common';

export class PrismaErrorHandler {
  static handle(error: any): { message: string; error: string | null } {
    if (error instanceof PrismaClientKnownRequestError) {
      switch (error.code) {
        case 'P2000':
          return {
            message: 'Error en la operación',
            error: 'El valor proporcionado para la columna es demasiado largo',
          };
        case 'P2001':
          return {
            message: 'Error en la operación',
            error: 'Registro no encontrado',
          };
        case 'P2002':
          return {
            message: 'Error en la operación',
            error: 'Violación de unicidad (El registro ya existe)',
          };
        case 'P2003':
          return {
            message: 'Error en la operación',
            error: 'Violación de la restricción de clave externa',
          };
        case 'P2004':
          return {
            message: 'Error en la operación',
            error: 'Violación de la restricción de chequeo',
          };
        case 'P2005':
          return {
            message: 'Error en la operación',
            error: 'Valor no válido',
          };
        case 'P2006':
          return {
            message: 'Error en la operación',
            error: 'Valor almacenado no válido',
          };
        case 'P2007':
          return {
            message: 'Error en la operación',
            error: 'Error inesperado de la base de datos',
          };
        case 'P2008':
          return {
            message: 'Error en la operación',
            error: 'Error en la consulta a la base de datos',
          };
        case 'P2009':
          return {
            message: 'Error en la operación',
            error: 'Error en la validación de datos',
          };
        default:
          return {
            message: 'Error en la operación',
            error: `Código de error no manejado: ${error.code}`,
          };
      }
    }
    //Handling errors not related to Prisma
    return {
      message: 'Error en la operación',
      error: error.message,
    };
  }
}
```

### Next steps

- Implement this specific error handling in all CRUD methods.
- Document specific error messages for future reference.
- Test extensively to ensure that all errors are handled correctly, and appropriate messages are returned. 

**Example of use in the `create` service:**

```ts
import { PrismaErrorHandler } from './prisma-error-handler';

async create( ) {
    try {
      // Logic to create
    } catch (error) {
      const handledError = PrismaErrorHandler.handle(error);
      return {
        message: handledError.message,
        error: handledError.error,
        data: null,
      };
    }
  }
```

## Benefits

- **Clear error messages:** Users will receive more specific and useful error messages.
- **Better diagnosis:** Developers will be able to identify and resolve problems more easily.
- **Improved user experience:** Users will have a better understanding of problems and how to address them.

**I hope this proposal will significantly improve error handling in our application, making both development and user experience more efficient.**

**Comment by DiegoSHS** [comment]
I find this concept a cool idea, but we can make this more efficient and compact
## Upgrades
we can upgrade the `PrismaErrorHandler` class by following this steps
- **Use Injectable decorator**
- **Replace switch with a key-value object**
- **Return directly the response object**
### Implementation
```ts
@Injectable()
export class PrismaErrorHandler {
    private readonly defaultMessages = {
        'P2000': 'El valor proporcionado para la columna es demasiado largo',
        'P2001': 'Registro no encontrado',
        'P2002': 'Violación de unicidad (El registro ya existe)',
        'P2003': 'Violación de la restricción de clave externa',
        'P2004': 'Violación de la restricción de chequeo',
        'P2005': 'Valor no válido',
        'P2006': 'Valor almacenado no válido',
        'P2007': 'Error inesperado de la base de datos',
        'P2008': 'Error en la consulta a la base de datos',
        'P2009': 'Error en la validación de datos',
    }
    private readonly notHandledError = 'Error no manejado';
    private readonly errorText = 'Error en la operación';
    /**
     * Method to handle errors from Prisma.
     * @param error Error to handle
     * @returns Error message
     */
    public handleError(error: any | PrismaClientKnownRequestError): { error: string; message: string, data: null } {
        if (error instanceof PrismaClientKnownRequestError) {
            const message: string = this.defaultMessages[error.code] || this.notHandledError;
            return {
                error: this.errorText,
                message,
                data: null,
            };
        }
        return {
            error: this.errorText,
            message: 'Error inesperado',
            data: null,
        };
    }
}
```
This is a similar aproach as the custom exception filter that we used in the past
### Using in a service
Injection of the `PrismaErrorHandler` as a provider
```ts
@Module({
  providers: [/*...other providers */, ExampleService, PrismaErrorHandler],
  controllers: [ExampleController],
})
export class ExampleModule {}
```
Using with the actual service
```ts
@Injectable()
export class ExampleService {
  constructor(private prisma: PrismaService, private readonly errorHandler: PrismaErrorHandler) {}
  async serviceMethod ( ): Promise<{ message: string | null; error: string | null; data: any | null }> {
    try {
      /* prisma logic */
    } catch (error) {
      return this.errorHandler.handleError(error);
    }
  }
}
```

**Comment by Ana202016** [comment]
Both proposals are good, since they allow us to be more specific when sending an error, on the other hand, the second proposal is more compact in terms of lines of code.  : )

**Comment by DiegoSHS** [comment]
# Override of existent functionalities
Some of the current functionalities do the same error handeling of some of the cases in this proposal of implementation
Existing functionalities that can be replaced or removed from the proposal:
- **Error for P2001:** The validator for foreign keys do this validation end avoid the error to be handled by prisma.
- **Error for P2002:** There's a validation for some insert operations that checks if the record exists
- **Error for P2003:** the validator for foreign keys do this validation and avoid the error to be handled by prisma
## Possible update for the proposal
I have two possible ways to do this
- Delete the existent validators for the specific cases for prisma errors and let this validator to provide all the messages
- Delete that cases of the prisma error handler and just update the implementation to send specific messages for every data resource
